test/amm/StableMath.test.ts

```diff
@@ -0,0 +1,127 @@
+import { AsyncFunc } from 'mocha';
+import { Contract } from 'ethers';
+import { ContractBase } from "../utils/ContractBase";
+import { BigNumberish, fp, num, isBigNumberish } from './Numbers';
+import { Decimal } from 'decimal.js';
+import {
+  amp,
+  calculateAnalyticalInvariantForTwoTokens,
+  calculateInvariant,
+  calcInGivenOut,
+  calcOutGivenIn,
+  calcBptOutGivenExactTokensIn,
+  calcTokenInGivenExactBptOut,
+  calcBptInGivenExactTokensOut,
+  calcTokenOutGivenExactBptIn,
+  calcTokensOutGivenExactBptIn,
+  calculateOneTokenSwapFeeAmount,
+  getTokenBalanceGivenInvariantAndAllOtherBalances
+} from './StableMath';
+
+function expectEqual(expected:BigNumberish|Decimal, actual:BigNumberish|Decimal) {
+  const _actual = num(expected);
+  const _expected = num(actual);
+  const maxErr = _expected * 0.001;
+  expect(_actual).to.be.within(_expected - maxErr, _expected + maxErr);
+  //console.log('OK actual=%s expected=%s', _actual, _expected);
+}
+
+async function expectEquals(fun1:AsyncFunc, fun2:Function, ...args:any[]): Promise<void> {
+  const actual = await fun1.apply(null, args);
+  const expected = fun2.apply(null, args);
+
+  if (isBigNumberish(actual)) {
+    expectEqual(actual, expected);
+  } else {
+    const actualArr = Object.keys(actual).map(key => actual[key]);
+    const expectArr = Object.keys(expected).map(key => expected[key]);
+    if (actualArr.length !== expectArr.length) {
+      throw new Error("Array length mismatch:"+actualArr.toString()+" != "+expectArr.toString());
+    }
+    for (let i = 0; i < actualArr.length; ++i) {
+      expectEqual(actualArr[i], expectArr[i]);
+    }
+  }
+}
+
+describe.only('StableMath', () =>
+{
+  let mock:Contract;
+
+  before(async () => {
+    mock = await ContractBase.deployContract("MockStableMath");
+  });
+
+  it('invariant', async () =>
+  {
+    const equal = (...args:any[]) => expectEquals(mock.invariant, calculateInvariant, ...args);
+    await equal(amp(100), /*balances*/[fp(10), fp(12)], true);
+  });
+
+  it('invariant equals analytical solution', async () =>
+  {
+    const equal = (...args:any[]) => expectEquals(mock.invariant, calculateAnalyticalInvariantForTwoTokens, ...args);
+    await equal(amp(100), /*balances*/[fp(10), fp(12)], true);
+  });
+
+  it('invariant reverts if it does not converge', async () =>
+  {
+    await expect(mock.invariant(amp(5000), [fp(0.00001), fp(1200000), fp(300)], true))
+          .to.be.revertedWith('StableMath no convergence');
+  });
+
+  it('inGivenOut', async () =>
+  {
+    const equal = (...args:any[]) => expectEquals(mock.inGivenOut, calcInGivenOut, ...args);
+    await equal(amp(100), /*balances*/[fp(10), fp(12)], /*indexIn*/0, /*indexOut*/1, /*tokenAmountOut*/fp(1));
+  });
+
+  it('outGivenIn', async () =>
+  {
+    const equal = (...args:any[]) => expectEquals(mock.outGivenIn, calcOutGivenIn, ...args);
+    await equal(amp(10), /*balances*/[fp(10), fp(11)], /*indexIn*/0, /*indexOut*/1, /*tokenAmountOut*/fp(1));
```

Comment:
I think this is token in
```suggestion
    await equal(amp(10), /*balances*/[fp(10), fp(11)], /*indexIn*/0, /*indexOut*/1, /*tokenAmountIn*/fp(1));
```

Reply:
Thanks, fixed 👍🏻